server/service/src/invoice/inventory_adjustment/insert.rs

```diff
@@ -0,0 +1,245 @@
+            .context(mock_store_a().id, "".to_string())
+            .unwrap();
+        let service = service_provider.invoice_service;
+
+        // Stockline doesn't exist
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: "x".to_string(),
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::StockLineDoesNotExist)
+        );
+
+        // Wrong store
+        context.store_id = mock_store_b().id;
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidStore)
+        );
+        context.store_id = mock_store_a().id;
+
+        // Missing reason
+        // assert_eq!(
+        //     service.insert_inventory_adjustment(
+        //         &context,
+        //         InsertInventoryAdjustment {
+        //             stock_line_id: mock_stock_line_a().id,
+        //             ..Default::default()
+        //         }
+        //     ),
+        //     Err(ServiceError::AdjustmentReasonNotProvided)
+        // );
+
+        // Invalid reason
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    inventory_adjustment_reason_id: Some("invalid".to_string()),
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidAdjustment)
+        );
+
+        // Reduce stock below zero
+        assert_eq!(
+            service.insert_inventory_adjustment(
+                &context,
+                InsertInventoryAdjustment {
+                    stock_line_id: mock_stock_line_a().id,
+                    adjustment: 0.0,
+                    ..Default::default()
+                }
+            ),
+            Err(ServiceError::InvalidAdjustment)
+        );
+
+        // Reduce stock below zero
+        // assert_eq!(
+        //     service.insert_inventory_adjustment(
+        //         &context,
+        //         InsertInventoryAdjustment {
+        //             stock_line_id: mock_stock_line_a().id,
+        //             adjustment_type:
+        //                 crate::invoice::inventory_adjustment::AdjustmentType::Reduction,
+        //             adjustment: 50.0,
+        //             ..Default::default()
+        //         }
+        //     ),
+        //     Err(ServiceError::StockLineReducedBelowZero(StockLine {
+        //         stock_line_row: mock_stock_line_a(),
+        //         ..Default::default()
+        //     }))
```

Comment:
The problem I have here is that the (pre-existing) StockLineReducedBelowZero Error requires a full `StockLine`, but we only have a `StockLineRow` available. At first I tried changing the enum definition (at line 36) to take a "StockLineRow" parameter, but that causes problems with the GraphQL error (in graphql insert, line 102) which requires a  full `StockLine` in its `from_domain()` method. I was a bit wary of changing *that* as well, as the cascade other other changes resulting from that was getting a bit messy. Any tips?

Reply:
Ah yep, I think the easiest way would be to use the StockLineRepository to query for the full stock line, so you can pass it in here. There is an example in stocktake_line: https://github.com/msupply-foundation/open-msupply/blob/develop/server/service/src/stocktake_line/insert.rs#L575